### PR TITLE
Specify group id for Slack whitelisted domains.

### PR DESCRIPTION
**Pull request by lasryaric**
## Description
Adding support for specifying a group ID associated with a Slack whitelisted domain.

We decided to go with this dirty solution of separating the domain name and the group id with `:` character in the database, given that this is only used by one customer and will be deprecated soon by the user of vaults.

## Deploy Plan
- Deploy connectors
- Run the backfill migration of this PR. There will be a few minutes where the whitelisted domains won't work but given that it's only for one customer, I think it's acceptable. We can do it over lunch time or tonight if needed.

### CODE REVIEW

**Review by flvndvd — COMMENTED** [review]
LGTM left some suggestions to improve code readability.

**Review by spolu — COMMENTED** [review]
The cli hasn't been adapted to this new format no?

**Comment by flvndvd** [comment]
> The cli hasn't been adapted to this new format no?

Yes you beed to update here as well:
https://github.com/dust-tt/dust/blob/0bc1d6499590952046b8c093b6d17be824b12292/connectors/src/connectors/slack/lib/cli.ts#L127

File: connectors/src/connectors/slack/lib/workspace_limits.ts [hunk 1]
```diff
@@ -249,28 +249,43 @@ async function isExternalUserAllowed(
...
+    .indexOf(userDomain);
+  if (whitelistedIndex === -1) {
+    return { authorized: false, groupIds: [] };
+  }
+  const whitelistedDomain = whitelistedDomains[whitelistedIndex]?.split(":")[0];
+  const whitelistedGroup = whitelistedDomains[whitelistedIndex]?.split(":")[1];
+
+  if (!whitelistedDomain || !whitelistedGroup) {
+    return { authorized: false, groupIds: [] };
+  }
```
[the remark this answers is not in the document]
Comment (lasryaric) [reply connectors/src/connectors/slack/lib/workspace_limits.ts:272]: It looks cleaner but this is pretty counter intuitive IMHO, it implies that there are potentially multiple matching entries. I will still commit the suggestion though :)

**Comment by lasryaric** [comment]
>The cli hasn't been adapted to this new format no?

@spolu Updated the CLI. Thanks.

@flvndvd , I also added a migration script.

File: front/migrations/20240903_backfill_slack_whitelisted_domains_groups.ts [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+import { Authenticator } from "@app/lib/auth";
+import { Workspace } from "@app/lib/models/workspace";
+import { GroupResource } from "@app/lib/resources/group_resource";
+import { makeSId } from "@app/lib/resources/string_ids";
+import { makeScript } from "@app/scripts/helpers";
+
+makeScript({}, async ({ execute }, logger) => {
+  const connectorDB = new Sequelize(
+    EnvironmentConfig.getEnvVariable("CONNECTORS_DATABASE_URI")
+  );
```
Comment (flvndvd) [inline front/migrations/20240903_backfill_slack_whitelisted_domains_groups.ts:13]: You can use `getConnectorReplicaDbConnection`.

File: front/migrations/20240903_backfill_slack_whitelisted_domains_groups.ts [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+  const slackConfigurations: {
+    whitelistedDomains: string[];
+    workspaceid: string;
+    configurationId: number;
+  }[] = await connectorDB.query(
+    `SELECT sc.id as "configurationId", sc."whitelistedDomains" as "whitelistedDomains", c."workspaceId" as workspaceId FROM slack_configurations sc INNER JOIN connectors c ON sc."connectorId" = c.id WHERE c.type = 'slack' and sc."whitelistedDomains" IS NOT NULL`,
```
Comment (flvndvd) [inline front/migrations/20240903_backfill_slack_whitelisted_domains_groups.ts:19]:
```suggestion
    `SELECT sc.id as "configurationId", sc."whitelistedDomains" as "whitelistedDomains", c."workspaceId" as "workspaceId" FROM slack_configurations AS sc INNER JOIN connectors AS c ON sc."connectorId" = c.id WHERE c.type = 'slack' and sc."whitelistedDomains" IS NOT NULL`,
```

File: front/migrations/20240903_backfill_slack_whitelisted_domains_groups.ts [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+      throw new Error(
+        `Failed to fetch global group for workspace ${sc.workspaceid}`
+      );
+    }
+    const workspace = await Workspace.findOne({
+      where: { sId: sc.workspaceid },
+    });
+    if (!workspace) {
+      throw new Error(`Workspace not found for workspaceId ${sc.workspaceid}`);
+    }
```
Comment (flvndvd) [inline front/migrations/20240903_backfill_slack_whitelisted_domains_groups.ts:38]: This is already done by the `Authenticator.internalAdminForWorkspace` so you can use `auth.getNonNullableOwner`.

File: front/migrations/20240903_backfill_slack_whitelisted_domains_groups.ts [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+
+    const groupId = makeSId("group", {
```
Comment (flvndvd) [inline front/migrations/20240903_backfill_slack_whitelisted_domains_groups.ts:40]:
```suggestion
    const groupId = GroupResource.modelIdToSId({
```

**Review by flvndvd — APPROVED** [review]
LGTM, left a few small nits to fix.

File: front/migrations/20240903_backfill_slack_whitelisted_domains_groups.ts (hunk 2 above)
Comment (lasryaric) [reply front/migrations/20240903_backfill_slack_whitelisted_domains_groups.ts:13]: I can't because I am updated the connectors DB at the end.

**Pull request merged**